Add playbook to secure shared gluster storage

ShyamsundarR opened the pull request:
Signed-off-by: ShyamsundarR <srangana@redhat.com>

ShyamsundarR committed:
Add playbook to secure shared gluster storage

Signed-off-by: ShyamsundarR <srangana@redhat.com>

ShyamsundarR commented:
Problem: Ansible play to configure shared storage TLS may have partial host list from the inventory, leading to incorrect auth.ssl-allow list or failure to restart the shared storage volume on all nodes.

For example, running the play as follows, will have all nodes that are a part of the volume being created, but not nodes that form the entire cluster.
ansible-ec2 -l tag_gluster_group_us_east_2a_c00_g00 ./playbooks/secure-shared-storage.yaml

Similarly a command as below, will have just a single node as a part of the play and not all that are part of the cluster.
ansible-ec2 -l us_east_1b_gluster_001 ./playbooks/secure-shared-storage.yaml

This leads to missing ansible_nodename variable for the remaining nodes and also inability for the play to unmount and remount the shared storage volume on all participating hosts.

Abilities present:
Irrespective of the host limit used in the play, it is possible to determine all hosts that are part of a group. This is under the assumption that we can determine the group that the hosts in the play is/are part of, and derive the list of hosts from the said group.

No-go's in terms of solution:
1. Inability to determine ansible_nodename if the host is not part of the play, and hence not part of the fact gathering process, makes it not possible to set the auth.ssl-allow list correctly, if all hosts of a group are not part of the play

2. Even if we determine all hosts in a group, if there are hosts outside the group but still part of the same gluster cluster, the determination and hence setting of the auth.ssl-allow will be erroneous.

Solution:
We need a tag (or equivalent) that groups hosts that are part of the same gluster cluster. This can be leveraged to understand if all hosts (that form the gluster cluster) are part of the play and hence it is feasible to run the play correctly, otherwise to abandon it.

Elaborating, assume every host has a variable `gluster-cluster-group` and this variable is a group name that contains all hosts that belongs to the said gluster cluster. It is possible to iterate and thus determine across play_hosts and the cluster group variable so that all hosts are part of the play and only these hosts are part of the play.

ShyamsundarR committed:
Updated shared storage role to strict check hosts list

Shared storage should be processed when play hosts are the same
as the gluster cluster hosts. Added checks to ensure this is
the case.

Further introduced a cluster_member-group to the inventory to
determine a list of hosts that belong to a cluster.

Signed-off-by: ShyamsundarR <srangana@redhat.com>

ShyamsundarR commented:
Addressed comment in https://github.com/JohnStrunk/oso-gluster-ansible/pull/32#issuecomment-376527689 as a part of the commit https://github.com/JohnStrunk/oso-gluster-ansible/pull/32/commits/4001a199ef68a14b9758d26621f0c58e94d4e1b4

ShyamsundarR commented:
@JohnStrunk If needed we would have to fix the following for this PR, else this is ready for review.

Assumption: A Gluster cluster can have more nodes than what a volume can (which was the case, but is not now)

Problem: The play to setup shared storage security cannot be run on hosts that form *a* volume, and has to be run on the nodes that form the cluster.

Solution:
- Define a group that contains all the nodes that form the cluster
- Change the cluster_member_group to that group value instead of the current values (this is the reason this variable is now defined in the inventory)
- Change the playbook secure-shared-storage and gluster_volume (part that invokes the gluster-secure-shared-storage) to run against the set of hosts from that group, across a list of such groups (rather than hosts that form a volume and across a list of volumes as it is presently)

ShyamsundarR commented:
> (after checking out state of code at commit c91f6f4 (pre-split of the cluster into volume units))
> BUT, that is wrong, the group is defined in inventory_groups, and the tag defines which volume the node serves, right?

Ok clarification :smile: , reorganizing the comment, as I wrote it, made this appear that the pre-split was wrong, it was meant to convey the explanation *I* provided above is wrong. IOW, my assumption that gluster_group tag meant it was a "cluster".

ShyamsundarR committed:
Corrected cluster_member_group to be the cluster than the volume group

Signed-off-by: ShyamsundarR <srangana@redhat.com>

ShyamsundarR commented:
> Couldn't these just be "g-us-east-1a-c00"?
Ack! corrected in commit https://github.com/JohnStrunk/oso-gluster-ansible/pull/32/commits/2152369c578ce067731d9daef06a8e1ec80962a4

JohnStrunk commented on playbooks/gluster-volume.yml:
```diff
@@ -12,3 +12,16 @@
            inventory_hostname in groups[gluster_volumes[volume].group]"
     loop_control:
       loop_var: volume
+
```
The contents here seem to match playbooks/secure-shared-storage.yaml. Should we just include the playbook to avoid code duplication, or am I missing something?

ShyamsundarR replied:
Yes, they are the same.

The deal though is, this fictional separation is wrong :)

We should ideally separate the plays that do cluster formation and setting up shared storage on that, with volume creation. Thus, when we actually get back to larger clusters but group based volumes, we end up doing the right thing. The initial separation of the include was for that, but as the latter is not done yet, this is fictional at present.

I want to open an issue to separate the plays here for gluster cluster setup versus gluster volume setup, as there is some cleanup needed here around that, and handle this then, is that fine?

ShyamsundarR replied:
Ok wrong answer above.

Yes they are the same, but I would prefer to not import playbooks in the playbook directory, and keep it to site.yml only.

JohnStrunk commented on playbooks/secure-shared-storage.yaml:
```diff
@@ -0,0 +1,15 @@
+# vim: set ts=2 sw=2 et :
```
To be consistent: file name *.yaml ==> *.yml

ShyamsundarR replied:
Ack!

ShyamsundarR replied:
Ack!

JohnStrunk commented on roles/gluster-secure-shared-storage/tasks/main.yml:
```diff
@@ -0,0 +1,39 @@
+- name: Configure shared storage access restrictions
+  include_tasks: "restrict_shared_storage.yml"
+  when: (ss_server_tls_status.stdout != "on") or (ss_client_tls_status.stdout != "on") or
+        (ss_tls_allow_status.stdout != groups[hostvars[inventory_hostname].cluster_member_group]|
```
Are there any issues around ordering of the host list? I don't know if groups[foo] returns a well defined order or not. I don't know that it would matter if we end up re-applying due to different sorting.
(Is this the false positive you mention?)

ShyamsundarR replied:
Good catch!

Two things,
- Order is from the group, which is formed using the tag, so, will the list order be the same across runs? I guess depends on how the tags are processed and added to the list, and hence I would assume not.
- Running the play as the allow rules are different, because of the order, does no harm at present.

Ideally (if I were to write python, or maybe go ;) ), it would be easier to convert the output of ssl-allow to a list and do the comparison. I am not going to trust my Ansible-fu at this point to pull that off (unless it is needed of course).

JohnStrunk commented on roles/gluster-secure-shared-storage/tasks/restrict_shared_storage.yml:
```diff
@@ -0,0 +1,58 @@
+    fail: msg="All hosts that form the gluster cluster are not part of the play, missing {{ gl_gluster_hosts }}"
+    when: gl_gluster_hosts|length != 0
+    vars:
+      gl_gluster_hosts: "{{ groups[hostvars[inventory_hostname].cluster_member_group] | difference(play_hosts) }}"
```
:1st_place_medal:

JohnStrunk commented on roles/gluster-secure-shared-storage/tasks/restrict_shared_storage.yml:
```diff
@@ -0,0 +1,58 @@
+    vars:
+      gl_this_play_hosts: "{{ play_hosts | difference(groups[hostvars[inventory_hostname].cluster_member_group]) }}"
+
+- name: Stop the gluster shared storage mounts
```
Does this work on hosts that were not part of the original 3-nodes that host shared_storage? I'm not sure the systemd unit gets enabled on the addl hosts. I've always done this via a straight unmount call (see gluster-volume where it sets up shared_storage)

ShyamsundarR replied:
The systemd unit comes with glusterfs-server package. So on systems that do not have this installed, it will not work. (file: /usr/lib/systemd/system/glusterfssharedstorage.service )

By additional hosts, you mean hosts that are part of the cluster but, are not part of any volume? Or, hosts that are not part of the cluster?

For the latter, this play wont even run, right?

For the former, glusterfs-server would be installed (as glusterd is part of the same package) and hence this would work as expected.

The systemd unit being enabled is an interesting observation. In my setup, before writing the ansible, when experimenting I did see the systemd unit stating the mount was inactive, and as I progressed through the various commands, and started using systemd to manage the service, things settled down. Not sure if this is a problem and why it occurs. (checked on the production systems and current status is as it was initially in my test system).

ShyamsundarR replied:
Based on observations, systemd may not be managing this instance of the service, and hence trusting that is not safe. Moving ahead with suggestions as per the review comment.

JohnStrunk reviewed (commented):
Overall looks good... just a couple questions.

ShyamsundarR committed:
Fixed umount and mount issues with systemd

In real clusters fstab gets to mount before systemd
hence using systemd to stop and start the mount will
not work.

Corrected other issues,
- .yml file extension

Signed-off-by: ShyamsundarR <srangana@redhat.com>

ShyamsundarR committed:
Merge branch 'master' into fix-tls-for-shared-storage

ShyamsundarR committed:
Removed starting shared storage mount post applying TLS settings

Signed-off-by: ShyamsundarR <srangana@redhat.com>
